contributing/architecture-eval-lifecycle.md

```diff
@@ -0,0 +1,354 @@
+# Architecture: Evaluations into Allocations
+
+The [Scheduling Concepts][] docs provide an overview of how the scheduling
+process works. This document is intended to go a bit deeper than those docs and
+walk thru the lifecycle of an Evaluation from job registration to running
+Allocations on the clients. The process can be broken into 4 parts:
+
+* [Job Registration](#job-registration)
+* [Scheduling](#scheduling)
+* [Deployment Watcher](#deployment-watcher)
+* [Client Allocs](#client-allocs)
+
+## Job Registration
+
+Creating or updating a Job is a _synchronous_ API operation. By the time the
+response has returned to the API consumer, the Job and an Evaluation for that Job
+have been written to the state store of the server nodes.
+
+* Note: parameterized or periodic batch jobs don't create an Evaluation at
+  registration, only once dispatched.
+* Note: The scheduler is very fast! This means that once the CLI gets a
+  response it can immediately start making queries to get information
+  about the next steps.
+* Note: This workflow is different for Multi-Region Deployments (found in Nomad
+  Enterprise). That will be documented separately.
+
+The diagram below shows this initial synchronous phase. Note here that there's
+no scheduling happening yet, so no Allocation (or Deployment) has been created.
+
+```mermaid
+sequenceDiagram
+
+    participant user as User
+    participant cli as CLI
+    participant httpAPI as HTTP API
+    participant leaderRpc as Leader RPC
+    participant leaderFSM as Leader State Store
+    participant workerFSM as Worker State Store
+
+    user ->> cli: nomad job run
+    activate cli
+    cli ->> httpAPI: Create Job API
+    httpAPI ->> leaderRpc: Job.Register
+
+    activate leaderRpc
+
+    leaderRpc ->> leaderFSM: write Job
+    activate leaderFSM
+    leaderFSM ->> workerFSM: replicate Job to followers
+
+    workerFSM -->> leaderFSM: ok
+    leaderFSM -->> leaderRpc: ok
```

Comment:
*Not a blocker!*

I think it might be beneficial to wrap this up in some single stage called `raftApply` or similar. It could then be reused in any graph we have a `raftApply` call.

The distinction is subtle, but I think worth making:

> `raftApply` returning without an error guarantees all servers _will be able to retrieve this object from their statestore at some point in the future._

The distinction is important because when a follower dequeues an evaluation it _may not be able to retrieve that evaluation from its statestore despite having applied the raft log!_ This is why the worker immediately ensures its statestore has the eval's WaitIndex: https://github.com/hashicorp/nomad/blob/v1.4.0-rc.1/nomad/worker.go#L415


Also I think making folks think about `raftApply` as an atomic operation is helpful. Sure there are network roundtrips in there to consider, but the whole point of `raftApply` is that if it succeeds a quorum of servers is guaranteed to have committed the log atomically.

Reply:
Had a sidebar with @schmichael about the notation to make that work, and we're going to collapse the workerFSM/leaderFSM into a single "raft" participant. That'll let us refer to the raft workflow outside this already huge diagram 😀

Reply:
Done